### PR TITLE
build!: update to pydantic 2

### PR DESCRIPTION
**Pull request by lengau**
- [ ] Have you followed the guidelines for contributing?
- [ ] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [ ] Have you successfully run `tox`?

-----
Failing tests are due to craft-grammar needing this merge: https://github.com/canonical/craft-grammar/pull/34

### CODE REVIEW

File: tests/unit/models/test_base.py [hunk 1]
```diff
@@ -51,8 +53,8 @@ def test_model_reference_slug_errors():
 
     expected = (
         "Bad testcraft.yaml content:\n"
-        "- bad value1 value (in field 'value1')\n"
-        "- bad value2 value (in field 'value2')"
+        "- value error, Bad value1 value (in field 'value1')\n"
```
Comment (tigarmo) [inline tests/unit/models/test_base.py:56]: I think as a follow-up we'll want to get rid of this "value error," prefix, it doesn't really add anything to the user

**Review by tigarmo — APPROVED** [review]
great!

File: craft_application/services/lifecycle.py [hunk 2]
```diff
@@ -187,7 +187,7 @@ def _init_lifecycle_manager(self) -> LifecycleManager:
             return LifecycleManager(
                 {"parts": self._project.parts},
                 application_name=self._app.name,
-                arch=convert_architecture_deb_to_platform(build_for),
```
[the remark this answers is not in the document]
Comment (lengau) [reply craft_application/services/lifecycle.py:190]: craft-parts switched to debian architecture names: https://github.com/canonical/craft-parts/pull/737

File: tests/unit/models/test_base.py (hunk 1 above)
Comment (lengau) [reply tests/unit/models/test_base.py:56]: I'm mixed about that. If we start using `PydanticCustomError` it'll probably be more useful (and we can subclass that to provide general error categories), and already it's helpful in distinguishing type errors from value errors.

File: craft_application/models/constraints.py [hunk 3]
```diff
@@ -14,78 +14,182 @@
...
-     * Must contain at least one letter
-     * May not start or end with a hyphen
-     * May not have two hyphens in a row
-    """
+def _validate_list_is_unique(value: list[T]) -> list[T]:
+    value_set = set(value)
+    if len(value_set) == len(value):
+        return value
+    dupes = [item for item, count in collections.Counter(value).items() if count > 1]
+    raise ValueError(f"Duplicate values in list: {dupes}")
```
[the remark this answers is not in the document]
Comment (lengau) [reply craft_application/models/constraints.py:33]: Good eye! This made me notice that I hadn't copied the tests over for this, too.

File: craft_application/models/constraints.py [hunk 4]
```diff
@@ -14,78 +14,182 @@
...
+
+
+def _get_validator_by_regex(
```
[the remark this answers is not in the document]
Comment (lengau) [reply craft_application/models/constraints.py:36]: Thanks! Should we consider making it something public then?

[the remark this answers is not in the document]
Comment (mr-cal) [reply craft_application/models/constraints.py:36]: Not sure yet - I'll find out soon if snapcraft will need to use it for something that already isn't in craft-application.

**Pull request merged**